Kernel/Devices/Audio/IntelHDA/Stream.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2023, Jelle Raaijmakers <jelle@gmta.nl>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/NonnullOwnPtr.h>
+#include <AK/OwnPtr.h>
+#include <Kernel/Devices/Audio/IntelHDA/Codec.h>
+#include <Kernel/IOWindow.h>
+#include <Kernel/Locking/SpinlockProtected.h>
+
+namespace Kernel::Audio::IntelHDA {
+
+class Stream {
+public:
+    static constexpr u32 cyclic_buffer_size_in_ms = 40;
+
+    bool running() const { return m_running; }
+    u8 bit_size() const { return m_pcm_bits; }
+    u8 channel_count() const { return m_channel_count; }
+    u32 sample_rate() const { return m_sample_rate; }
+
+    void start();
+    ErrorOr<void> stop();
+
+    ErrorOr<void> set_format(u32 sample_rate, u8 pcm_bits, u8 number_of_channels);
+
+protected:
+    // We always need 2 filled buffers, plus an additional one to prevent buffer underrun
+    static constexpr u8 minimum_number_of_buffers = 3;
+    // Arbitrary (unspecced) delay value of 1 link frame duration
+    static constexpr size_t register_change_delay_in_frames = 1;
+
+    // 3.3: High Definition Audio Controller Register Set - streams
+    enum StreamRegisterOffset : u8 {
+        Control = 0x00,
+        Status = 0x03,
+        LinkPosition = 0x04,
+        CyclicBufferLength = 0x08,
+        LastValidIndex = 0x0c,
+        Format = 0x12,
+        BDLLowerBaseAddress = 0x18,
+        BDLUpperBaseAddress = 0x1c,
+    };
+
+    // 3.3.35: Input/Output/Bidirectional Stream Descriptor Control
+    enum StreamControlFlag : u32 {
+        StreamReset = 1 << 0,
+        StreamRun = 1 << 1,
+    };
+
+    Stream(NonnullOwnPtr<IOWindow> stream_io_window)
+        : m_stream_io_window(move(stream_io_window))
+    {
+    }
+
+    ~Stream();
+
+    u32 read_control();
+    void write_control(u32);
+
+    ErrorOr<void> initialize_buffer();
+    ErrorOr<void> reset();
+
+    NonnullOwnPtr<IOWindow> m_stream_io_window;
+    OwnPtr<Memory::Region> m_buffer_descriptor_list;
+    SpinlockProtected<OwnPtr<Memory::Region>, LockRank::None> m_buffers;
+    size_t m_buffer_position { 0 };
+    bool m_running { false };
+    u32 m_sample_rate { 0 };
+    u8 m_pcm_bits { 0 };
+    u8 m_channel_count { 0 };
+};
+
+class OutputStream : virtual public Stream {
+public:
+    static ErrorOr<NonnullOwnPtr<OutputStream>> create(NonnullOwnPtr<IOWindow> stream_io_window)
+    {
+        return adopt_nonnull_own_or_enomem(new (nothrow) OutputStream(move(stream_io_window)));
+    }
+
+    ErrorOr<size_t> write(UserOrKernelBuffer const&, size_t);
+
+private:
+    OutputStream(NonnullOwnPtr<IOWindow> stream_io_window)
+        : Stream(move(stream_io_window))
+    {
+    }
+};
+
+class InputStream : virtual public Stream {
+    // FIXME: implement InputStream
+};
+
+class BidirectionalStream : public OutputStream
+    , public InputStream {
```

Comment:
I hope that this means any bidi stream will at least "just" work as an output stream? Otherwise I don't think it's sensible to have this inheritance at this point.

Reply:
They can, but there is some additional logic to them that I don't feel like explaining in `FIXME`s right now. This inheritance was only there to point future devs to the intended setup I had in mind, but I've removed it altogether and left a single `FIXME` since I don't want to get in the way of better designs.